pkg/kubelet/kubelet.go

```diff
@@ -349,12 +369,37 @@ func (self *podWorkers) Run(podFullName string, action func()) {
 	}()
 }
 
-func makeEnvironmentVariables(container *api.Container) []string {
+func (kl *Kubelet) makeEnvironmentVariables(container *api.Container) ([]string, error) {
 	var result []string
+	// Add service environment variables.  Note:  These are added to the docker.Config, but are
+	// not included in the checksum computed by dockertools.BuildDockerName(...).  That way,
+	// we can still determine whether an api.Container is already running by its hash.
+	// (We don't want to restart a container just because some service changed.)
+	//
+	// TODO: limit env vars to same namespace services, plus kubernetes{-ro}. (Issue #3293)
+	//
+	// Note that there is a race between Kubelet seeing the pod and kubelet seeing the service.
+	// To avoid this users can: (1) wait between starting a service and starting; or (2) detect
+	// missing service env var and exit and be restarted; or (3) use DNS instead of env vars
+	// and keep trying to resolve the DNS name of the service (recommended).
+	serviceEnv, err := kl.getServiceEnvVarMap()
+	if err != nil {
+		return result, err
+	}
 	for _, value := range container.Env {
+		// The code is in transition from using etcd+BoundPods to apiserver+Pods.
+		// So, the master may set service env vars, or kubelet may.  In case both are doing
+		// it, we delete the key from the kubelet-generated ones so we don't have duplicate
+		// env vars.
+		// TODO: remove this net line once all platforms use apiserver+Pods.
+		delete(serviceEnv, value.Name)
```

Comment:
Not all cloud providers have been converted to pass an apiserver argument
to the kubelet.  Until that is done, we can't have kubelet setting env vars
for all cloud providers.  So, apiserver needs to set env vars on all cloud
providers, or we need a flag to control apiserver behavior.  I thought the
former was easier.  But it requires deduplication.

On Mon, Jan 12, 2015 at 8:06 AM, Paul Morie <notifications@github.com>
wrote:

> In pkg/kubelet/kubelet.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3376#discussion-diff-22799515>
> :
>
> >  	for _, value := range container.Env {
> > +		// The code is in transition from using etcd+BoundPods to apiserver+Pods.
> > +		// So, the master may set service env vars, or kubelet may.  In case both are doing
> > +		// it, we delete the key from the kubelet-generated ones so we don't have duplicate
> > +		// env vars.
> > +		// TODO: remove this net line once all platforms use apiserver+Pods.
> > +		delete(serviceEnv, value.Name)
>
> @erictune <https://github.com/erictune> I think we might have a mismatch.
> I was going to change the scope of my PR to be to set all the service env
> vars in the kubelet. In that case, I don't think any deduplication will be
> necessary. Agree?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3376/files#r22799515>
> .
>